### ISSUE TITLE
Narrowing

### ISSUE THREAD

**Issue by alensiljak** [orgzly-community/orgzly-android#9]
One thing I really miss in Orgzly is narrowing. 
Issue in the original repo:
https://github.com/orgzly/orgzly-android/issues/21

**Comment by kidou343** [comment]
I think this could be really nice as a swipe-gesture thing.

**Comment by alensiljak** [comment]
It could be a separate action, like "narrow" and "expand" or whatever is the OrgMode terminology.

**Comment by amberin** [comment]
Yes, it could be a swipe action. But we could easily make it configurable, so that narrowing can be set as the default click action.

I'm mostly just wondering what the UI should look like to clarify what's going on, and make it easy to "zoom out" again.

**Comment by alensiljak** [comment]
I would try to make it as similar to OrgMode as possible. I.E. some indication that we are in narrowing mode, perhaps a line on top of the content window.
Apparently, OrgMode just shows Narrow in the mode indicator. Then the user can widen out of it to the full context.

**Comment by amberin** [comment]
> Apparently, OrgMode just shows Narrow in the mode indicator. Then the user can widen out of it to the full context.

Yeah, that's the thing, it's very discreet in OrgMode, and we have no status line. So we need to find something that's not too ugly, but fairly clear, I think. Personally, I lack the UI and Android skills to easily do mock-ups.

**Comment by alensiljak** [comment]
It can be a simple box with text above the list or just the presence of Widen option in the context menu. It is not that important. The assumption is that the one using the application knows whether they are in narrowing mode or not. 
And the default is to reset the mode on opening a note. If one follows a link and then navigates back, however, then we should keep it in the same mode.

**Comment by ShuoShen** [comment]
> so that narrowing can be set as the default click action

Slightly tangential to the main topic: this makes more sense because it's less intrusive than showing the edit view. By default people want to zoom into a piece of content, be it a post, picture, or in our case a note. Then following the zoom-in, some want to edit the content, and for that they can use a button somewhere a swipe. 

Speaking of design, I'm no expert. @alensiljak's suggestion works. 

If we want to explore alternative ideas, we can try something like this, where the narrowed note is highlighted, and there's a small portion of elements from its parent displayed in a blurred background. Scrolling only impacts the narrowed note. Tapping or double tapping the parent area gets us back to the wide view. 

![Image](https://github.com/user-attachments/assets/2cedcaec-ed05-4ac2-8283-6fac6dce2acb)

**Comment by ShuoShen** [comment]
> It can be a simple box with text above the list, with a link to "widen", or even just the presence of the Widen option in the context menu. It is not that important. The assumption is that the one using the application knows whether they are in narrowing mode or not.

This sounds pretty reasonable

> And the default is to reset the mode on opening a note.

Can you help me understand what are the modes and how does it reset? Does it mean clicking on a note would narrow to a subtree instead of showing an edit view? This feels a more natural behavior to me.

**Comment by aeshna-cyanea** [comment]
I quite like how logseq (an outliner based partially on org-mode) android app handles it- they have little buttons on the right edge of the screen for every node with children for expand/collapse, and the bullet points are clickable for narrowing.

<img src="https://github.com/user-attachments/assets/8aeee34d-ed9f-4b41-9e5a-0bb0890a7892" width="300px" />

**Comment by alensiljak** [comment]
OK, this is the reply. I'll fix the original post. Oh, man, sorry about that. It's fairly bad doing this on the phone, obviously. Instead of "quote reply" I got "Edit".
---

>By default people want to zoom into a piece of content

That would be an interesting option. At the moment, there is a setting for choosing between Select and Edit as two actions on tap. This could expand into a choice of default actions on tap and long-press. Would be worth experimenting to see what action works well as the default.

**Comment by alensiljak** [comment]
> Can you help me understand what are the modes and how does it reset? Does it mean clicking on a note would narrow to a subtree instead of showing an edit view? This feels a more natural behavior to me.

The modes are, obviously "narrow" and "wide", in a sense. The default is the wide mode, where the whole note is visible. Narrowing makes the chosen note the root (not top). I mean, I'm not trying to reinvent the wheel. The explanation is here:

https://www.gnu.org/software/emacs/manual/html_node/emacs/Narrowing.html

>     1. Clicking on a note would open the "narrowed view".
> I believe this workflow makes more sense because it provides a more coherent viewing experience, and viewing is the most common mode. Whereas the current workflow of "click-to-edit" is intrusive. Obviously this would add friction to edits, and take some time to get used to. But I feel it's a reasonable tradeoff, given how much infrequent a user actually needs to edit.

We should not change any default behaviour. Apps that do that are just annoying. There is already an option in the Settings to choose what one wants as the default action and Narrowing should simply be one more option there.

> [@alensiljak](https://github.com/alensiljak), lol I think you edited my previous comment instead of replying to it. I thought I was having deja vu when writing this comment :P

Yes, sorry, I was doing it on the phone and got Edit instead of Quote Reply, I believe. Didn't know it was possible to edit others' posts. The rest of the behaviour is identical so I did not even realize.

**Comment by ShuoShen** [comment]
> I quite like how the logseq (an outliner based partially on org-mode) android app handles it- they have little buttons on the right edge of the screen for every node with children for expand/collapse, and the bullet points are clickable for narrowing.
> 
> <img alt="" width="300px" src="https://private-user-images.githubusercontent.com/186870771/439081914-8aeee34d-ed9f-4b41-9e5a-0bb0890a7892.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYwNDk5ODcsIm5iZiI6MTc0NjA0OTY4NywicGF0aCI6Ii8xODY4NzA3NzEvNDM5MDgxOTE0LThhZWVlMzRkLWVkOWYtNGI0MS05ZTVhLTBiYjA4OTBhNzg5Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQzMFQyMTQ4MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMTkxZWJmOTAwYjI5NDRiMmNlZmEzYWJjNTE3YmE5MGJlZjcxNzAyNDA4NDcwNDQ5MTAyMjAzMDEwOGU0MmI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.TL0mp37kllLJkMCvHeaNAR5z46gCQ-jUq9pi9AlayWI">


@aeshna-cyanea, thanks for sharing. Logseq's navigation flow does feel more natural than orgzly. One big difference between logseq and orgzly is that logseq doesn't seem to have a separate edit-mode, and all edits are done inline at the same time as viewing. For orgzly, inline edit will be more disruptive because of all the timestamps and properties that needs specially input mode. 

That said, we could probably borrow its UX for narrow, i.e. clicking the dot goes to narrow. The narrowed view shows a path to the note. 

<img width="350" alt="Image" src="https://github.com/user-attachments/assets/8c7b7cdd-8b2d-4f79-9fbf-109278aeb5fe" />

**Comment by ShuoShen** [comment]
> The modes are, obviously "narrow" and "wide", in a sense. The default is the wide mode, where the whole note is visible. Narrowing makes the chosen note the root (not top). I mean, I'm not trying to reinvent the wheel. The explanation is here:

Makes sense. Logseq's narrow view makes a lot of sense. I don't see why we can't just borrow their UX, i.e. a click on the dot enters the narrowed tree, and the narrowed view shows a path on top for navigation. 

> We should not change any default behaviour. Apps that do that are just annoying. 

That’s fair. Changing defaults does come with tradeoffs, and I totally agree it shouldn’t be done lightly.

That said, sometimes a default can genuinely feel misaligned with the most common or expected behavior, and it might be worth at least discussing whether that’s the case here.

Out of curiosity: do we all personally feel that the current default note-click action (edit view) works well for most users? Or would you also prefer something like Narrowing to be the default, even if it remains configurable?

> There is already an option in the Settings to choose what one wants as the default action and Narrowing should simply be one more option there.

I searched for a bit but couldn't find anything along those lines. Could you please share a screenshot of the pref name? Thanks.

> Yes, sorry, I was doing it on the phone and got Edit instead of Quote Reply, I believe. Didn't know it was possible to edit others' posts. The rest of the behaviour is identical so I did not even realize.

No worries. I just found that hilarious that github allows people to do that.

**Comment by alensiljak** [comment]
I like the navigation bar (breadcrumbs). It would need to be scrollable for deep notes and cut off long titles. It should also allow switching context by tapping on any of the parents. Then it would be a nice way to widen out, as well.

Tapping the dot would work but I find targeting the dot with a finger is fairly unintuitive and very frustrating endeavor. Something more explicit, like context or toolbar menu seems more appealing to me.

The setting in Orgzly is for swapping the tap and long-tap actions, open and select, respectively. That means those two are the main actions on the note, to open (view/edit) and move around. To me, open is the most useful action because I usually want to see the contents of the note. Narrowing is not something I'd expect to perform as often, in comparison. I also think that changing this default behavior would be a really bad idea, in general.
If anybody felt it was a problem, I'm sure there would be multiple requests to use some more sensible action instead.

**Comment by ShuoShen** [comment]
> To me, open is the most useful action because I usually want to see the contents of the note. 

Exactly. 

> Narrowing is not something I'd expect to perform as often, in comparison. 

Wouldn't narrowing be more helpful in viewing the content compared to today's open-to-edit?

**Comment by alensiljak** [comment]
> Wouldn't narrowing be more helpful in viewing the content compared to today's open-to-edit?

I don't see how narrowing helps in that sense, since you can already expand the note and see its children even before the view has been narrowed to it.
Narrowing would not display the content of the note.
But, as suggested, having an option to choose the default behavior would definitely be a good idea and is probably worth improving separately.

**Comment by ShuoShen** [comment]
> I don't see how narrowing helps in that sense, since you can already expand the note and see its children even before the view has been narrowed to it.

It will help. Today, it's hard to see a subnote when we have many different levels on a mobile phone. A note at a deep level will be barely readable because of the indentation. This can be solved by the "narrowing" (or "focus") to the subnote, such that it gets the entire width of the screen. 

> Narrowing would not display the content of the note.

I might be mis-understanding the concept then. What do you propose the narrowing should do? The org-narrow-to-subtree is defined as "Narrow buffer to current subtree" (https://orgmode.org/manual/Structure-Editing.html#index-C_002dx-n-s). That would solve the deep-note-viewing problem I mentioned above.

**Comment by amberin** [comment]
> Narrowing would not display the content of the note.

This is a very important clarification. Orgzly has the concept of "preface" in its books, which is all content before the first heading node. If our narrowed view should be similar to the book view, then we need to decide how to handle note content before the first sub-heading. I believe it should become the "preface" in Orgzly terms. This means the user can choose in preferences whether it's shown, hidden or partly shown. The default is to show all of it. In that case, @alensiljak's statement would be untrue - the note contents before the next sub-heading would be shown.

**Comment by amberin** [comment]
> Tapping the dot would work but I find targeting the dot with a finger is fairly unintuitive and very frustrating endeavor. Something more explicit, like context or toolbar menu seems more appealing to me.

Yes, the dot currently has two accessibility issues: the hit area is way too small, and the visual difference between folded vs empty is also too small. But these issues need to be addressed no matter what, so image the dot being easy to use. :)

**Comment by amberin** [comment]
Regarding changing the default behavior, I feel we could save that discussion for later. Let's start by not touching the current defaults and adding s different opt-in behavior. Once that feels really good, we can weigh pros vs cons of changing the defaults.

**Comment by ShuoShen** [comment]
> This is a very important clarification. Orgzly has the concept of "preface" in its books, which is all content before the first heading node. If our narrowed view should be similar to the book view, then we need to decide how to handle note content before the first sub-heading. I believe it should become the current "preface" in Orgzly terms. This means the user can (already now) choose in preferences whether it's shown, hidden or partly shown. The default is to show all of it. In that case, @alensiljak's statement would be untrue - the note contents before the next sub-heading would be shown.

@amberin Yes, I agree that retaining the preface behavior in the narrowed view makes sense, especially if the narrowed view is roughly aligned with book view. 

@amberin @alensiljak Just to confirm we’re on the same page:
	1. It sounds like we’re aligned on the “narrow” view behaving similarly to org-narrow-to-subtree, and UI looks roughly the same as book view with the narrowed note displayed with its "preface" (according to preface settings) along with sub notes.
	2. From an implementation standpoint, we can proceed with narrowing via clicking the dot, despite it's not the easiest button to click on. We can further improve the UX from there.
	3. The behavior of clicking the note title and handling the note preface can be treated as a follow-up discussion — particularly when we’re ready to explore default behaviors.

Let me know if I missed or misunderstood anything here.

**Comment by aterentic** [comment]
Ive tried to add narrowing as swipe left button ( -> ):
<img width="540" height="501" alt="Image" src="https://github.com/user-attachments/assets/4a102788-f122-427e-a563-309c40d597fb" />

and widening as top toolbar button ( <- ):
<img width="540" height="841" alt="Image" src="https://github.com/user-attachments/assets/3c785bcb-a6bf-4d0b-b531-9f8a23342f65" />

and its kind of nice, its not intrusive but its easily accessible, if needed. I didn't know that narrowing exists before reading this issue (🤦 ), so I hope I didn't miss the point. If thats the case, I can create PR.

**Comment by alensiljak** [comment]
Cool! Nice effort.
I like the Narrow button but I'm not sure that cramming more buttons into the toolbar is a good idea for Widen. I agree that it should be obvious that we are narrowed (focused, in other apps) but it would be good to have some other way of indicating that and exiting / widening.
Actually, that button would be quite OK if the "collapse/expand all" button would permanently move to the toolbar menu. I wonder if that is such a frequent action that it deserves a place in the main toolbar.
Very interesting work. I'd love to try it out. @amberin, what do you think?

Edit: there is also the swipe-left gesture, that shows structural actions on a node. Maybe that would be an alternative candidate for the button placement?

**Comment by amberin** [comment]
Awesome to see some progress here!

I'm probably missing something, but I don't really understand the difference between the two pictures, except that the nodes above the focused one are hidden? In my mind, the main point of narrowing is to reduce the number of levels, so that there is more room on the screen when viewing/showing a deeply nested node.

I'm also not sure if I would want to have the notebook "preface" visible when narrowing on a node. I'm thinking a major use case is to show a subtree to someone else, while hiding everything outside of the subtree.

Regarding the buttons, I don't have strong opinions. Swiping to narrow seems perfect to me. I think the back button should be clearly visible, but I agree with @alensiljak that the top menu becomes a little crowded. Personally, I use the fold/unfold button quite frequently. But during narrowing, maybe it can move to the overflow menu if space is tight? (Maybe use the "ifRoom" value of of app:showAsAction in the XML layouts.)

**Comment by alensiljak** [comment]
Right. I assumed the narrowing is doing what narrowing means. ;)
In case it is not, here is some clarification.
The original doc is [here](https://orgmode.org/manual/Structure-Editing.html) but it does not explain much.
The `org-narrow-to-subtree` function narrows the current buffer to make only the current subtree visible. Everything outside the subtree (including higher-level headings) becomes temporarily inaccessible.
There is also a less common `org-narrow-to-block`, which narrows the buffer to the current block (like a src block or example block).
It is not that complicated. The selected node becomes the root node of the buffer. This mainly affects the display of the child nodes in the tree. The Content of the selected node, on the other hand, can be modified in the regular Edit view.
I assume the screenshots show just the buttons, not the whole functionality?

**Comment by aterentic** [comment]
Actually its whole functionality, but not as great examples of it 🫤 First screenshot shows where would narrow button be (if configured), and second one shows the result, all other notes are hidden, except that subtree. But thats not easily visible on screenshots, obviously :) I have fixed issue with indentation (narrowed note has no indentation rn, removed preface when narrowed, and set fold//unfold to move to overflow menu). 

<img width="540" alt="Image" src="https://github.com/user-attachments/assets/26127224-7f80-4973-9f9c-995036648afe" />

I have to check if move note option should be disabled for root note when narrowed, and then Ill create PR so others can tryout. Btw, another surprise was that narrowing exists outside of org mode, lol.

**Comment by alensiljak** [comment]
That looks great!

> if move note option should be disabled

I have no preferences if this needs to be removed but it probably makes sense. There is no need to react to the root node swipes at all. It would probably be good just to show its title in a label, not being part of the tree at all.

> another surprise was that narrowing exists outside of org mode

It's becoming more common in Markdown editors / "productivity tools", as they are called nowadays. I think many tools, such as Notion, have this. I think good concepts from various tools are being consolidated constantly and new concepts emerge all the time. Although, nothing really new but reorganized, let's say. :)
I'm currently looking at AnyType. They have several new concepts (like CRDTs, for example), which are pretty cool in the internet age. Hopefully this local-first infrastructure will improve and we could simply utilize something like that in the future.

**Comment by aterentic** [comment]
Thanks for reference, Ill check AnyType, always thought I will have a need for CRDTs, but its still easier to tell my wife to sync notes manually 😆 I, have created PR, please check once you have time.
